Perf test dashboard setup (#212)
dapr/test-infra: deploy/aks/monitoring/perf/README.md

## Performance Test Dashboard Setup

This document serves as a guideline for setting up a performance test dashboard using Azure managed Prometheus and Grafana. We will be creating an AKS (Azure Kubernetes Service) cluster and installing Prometheus Pushgateway and Azure Monitoring Agent (AMA) to scrape performance test metrics and push them to Azure managed Prometheus. Additionally, we will set up an ingress controller with authentication to act as a proxy for Prometheus Pushgateway. Finally, we will create a Grafana dashboard by importing a predefined JSON model.

Below are step-by-step instructions to set up a performance test dashboard using Azure managed Prometheus and Grafana. Follow these steps to configure your performance monitoring environment.

#### Step 1: Clone the test-infra repo

```bash
git clone https://github.com/dapr/test-infra.git
cd deploy/aks/monitoring/perf
```

#### Step 2: Login to Azure

```bash

> az login

Reviewer: Something that I would add (because I have been bitten by this a number of times) is a command to ensure you are operating at the desired subscription. Performing log in does not lock commands to a particular subscription. This is less of an issue if you are logging in to an account that only has a single subscription but it becomes troublesome if you account (like ours) has access to multiple directories (logging in does not lock you to a directory) and if that directory itself hosts multiple subscriptions.

Something like the following would do the trick. I will send a PR changing the main README in this repo to do something to that effect

```bash
export SUBSCRIPTION_ID=<SUBSCRIPTION UUID TO BE USED FOR THE COMMANDS BELLOW>

# ...

az account set --subscription "${SUBSCRIPTION_ID}"
```

```

### Step 3: Set resource group name, location and use the same namespace (dapr-perf-metrics)

```bash

> DAPR_PERF_RG={reosuceGroup}
> DAPR_PERF_LOCATION={region}

Reviewer: Nitpicking I would use `<something>` instead of '{something}' to document something to be changed.

First because it forces the user to replace that: `{something}` is a valid bash string that parses without an error whereas '<something is not>':

```
$ echo <a>
bash: syntax error near unexpected token `newline'
$ echo '{a}'
{a}
````

Second because, to me at least, `{something}` makes me believe that you just forgot a `$` before that, and that you instead meant to refer to a env. variable named `something` as in `${something}`

Also, you have a typo in `reosuceGroup`

```suggestion
DAPR_PERF_RG=<resource group to be used>
DAPR_PERF_LOCATION=<insert region>
SUBSCRIPTION_ID=<subscription UUID to be used>
```

DAPR_PERF_METRICS_NAMESPACE=dapr-perf-metrics
```

#### Step 4: Create Resource Group
```bash
az group create --name $DAPR_PERF_RG --location $DAPR_PERF_LOCATION
```

#### Step 5: Execute main.bicep and provide AKS cluster name on prompt

```bash
az deployment group create --resource-group $DAPR_PERF_RG --template-file main.bicep
```

#### Step 6: Merge Newly Created Cluster Username and Password

```bash
az aks get-credentials --resource-group $DAPR_PERF_RG --name {clusterName}
```

#### Step 7: Switch AKS Cluster Context

```bash

> kubectl config use-context {clusterName}

Reviewer: Is `{clusterName}` an env. var that was not defined or something that the user was supposed to manually replace in these 2 commands?

It might be better to just define this as an env. var back in step 3 or before and use `${CLUSTER_NAME}` here instead.

```

#### Step 8: Install Prometheus Pushgateway

```bash
helm repo add prometheus-community https://prometheus-community.github.io/helm-charts
helm repo update
helm install --namespace $DAPR_PERF_METRICS_NAMESPACE prometheus-pushgateway prometheus-community/prometheus-pushgateway
```

> #### Step 9: Install Ingress Controller. 

Reviewer: I am wondering: isn't there a AKS setting or construct that could replace this?

Author: I looked up in the docs and over the internet, didn't see the option. If we find the option, we will update the bicept script. For now, we can go with this.

Follow this [linke](https://learn.microsoft.com/en-us/azure/aks/ingress-basic?tabs=azure-cli#basic-configuration) for more details on setting up nginx ingress controller.

```bash
helm repo add ingress-nginx https://kubernetes.github.io/ingress-nginx
helm repo update

helm install ingress-nginx ingress-nginx/ingress-nginx \
  --namespace $DAPR_PERF_METRICS_NAMESPACE \
  --set controller.service.annotations."service\.beta\.kubernetes\.io/azure-load-balancer-health-probe-request-path"=/healthz
```

#### Step 10: Create Username and Password for Authentication

To create a basic authentication [username and password](https://kubernetes.github.io/ingress-nginx/examples/auth/basic/), use the following command, which will create an auth file and prompt you to provide a username and password.

```bash
htpasswd -c auth {userName}
```

#### Step 11: Create a Secret in Kubernetes

```bash

> k create secret generic basic-auth --from-file=auth -n dapr-perf-metrics

Reviewer: Is `k` an alias you have set up? It would be better to have it spelled out to its full form instead.

```

#### Step 12: Create Ingress for Prometheus Pushgateway

```bash
k apply -f ./prometheus-pushgateway-ingress.yaml
```

#### Step 13: Create a Config Map for Service Discovery for AMA Agent

```bash
k apply -f ./prometheus-pushgateway-configmap.yaml
```

#### Step 14: Add user to grafana

- Go to grafana resource in Azure portal
- Select Access control (IAM) on left menu
- Click on Add role assignment
- Select suitable role for the user, and click Next
- In the Member tab, click on `+ Select Member` and type their email in search box
- Select user and click on `Review + assign`

#### Step 15: Create Grafana Dashboard

> Grab the granfa link from azure portal and create a Grafana dashboard by importing the [JSON model](https://github.com/dapr/dapr/blob/master/tests/grafana/grafana-perf-test-dashboard.json). Ensure to update the `uid` in the JSON to match your configuration.

Reviewer: UID of what (field)?

Author: updated the message.

```bash
az login
export SUBSCRIPTION_ID=<SUBSCRIPTION UUID TO BE USED FOR THE COMMANDS BELLOW>
az account set --subscription "${SUBSCRIPTION_ID}"
```

### Step 3: Set resource group name, location, cluster name and use the same namespace (dapr-perf-metrics)

```bash
DAPR_PERF_RG=<resource group to be used>  
DAPR_PERF_LOCATION=<insert region>  
CLUSTER_NAME=<cluster name>
DAPR_PERF_METRICS_NAMESPACE=dapr-perf-metrics
```

#### Step 4: Create Resource Group
```bash
az group create --name $DAPR_PERF_RG --location $DAPR_PERF_LOCATION
```

#### Step 5: Execute main.bicep and provide AKS cluster name on prompt

```bash
az deployment group create --resource-group $DAPR_PERF_RG --template-file main.bicep --parameters clusterName="${CLUSTER_NAME}"
```

#### Step 6: Merge Newly Created Cluster Username and Password

```bash
az aks get-credentials --resource-group $DAPR_PERF_RG --name "${CLUSTER_NAME}"
```

#### Step 7: Switch AKS Cluster Context

```bash
kubectl config use-context "${CLUSTER_NAME}"
```

#### Step 8: Install Prometheus Pushgateway

```bash
helm repo add prometheus-community https://prometheus-community.github.io/helm-charts
helm repo update

> helm install --namespace $DAPR_PERF_METRICS_NAMESPACE prometheus-pushgateway prometheus-community/prometheus-pushgateway

Reviewer: I don't think you created this namespace before.  How about letting helm take care of this. And how about making this command idempotent? 

```suggestion
helm upgrade --install \
    prometheus-pushgateway prometheus-community/prometheus-pushgateway \
    --namespace $DAPR_PERF_METRICS_NAMESPACE \
    --create-namespace \
    --wait
```

> helm install --namespace $DAPR_PERF_METRICS_NAMESPACE prometheus-pushgateway prometheus-community/prometheus-pushgateway

Reviewer: I don't think you created this namespace before.  How about letting helm take care of this. And how about making this command idempotent? 

```suggestion
helm upgrade --install \
    prometheus-pushgateway prometheus-community/prometheus-pushgateway \
    --namespace $DAPR_PERF_METRICS_NAMESPACE \
    --create-namespace \
    --wait
```

Author: It is created as part of bicep template.

```

#### Step 9: Install Ingress Controller. 

> Follow this [linke](https://learn.microsoft.com/en-us/azure/aks/ingress-basic?tabs=azure-cli#basic-configuration) for more details on setting up nginx ingress controller.

Reviewer: ```suggestion
Follow this [link](https://learn.microsoft.com/en-us/azure/aks/ingress-basic?tabs=azure-cli#basic-configuration) for more details on setting up nginx ingress controller.
```

```bash
helm repo add ingress-nginx https://kubernetes.github.io/ingress-nginx
helm repo update

> helm install ingress-nginx ingress-nginx/ingress-nginx \

Reviewer: ```suggestion
helm upgrade --install \
    ingress-nginx ingress-nginx/ingress-nginx \
```

```

#### Step 8: Install Prometheus Pushgateway

```bash
helm repo add prometheus-community https://prometheus-community.github.io/helm-charts
helm repo update
helm upgrade --install \
    prometheus-pushgateway prometheus-community/prometheus-pushgateway \
    --namespace $DAPR_PERF_METRICS_NAMESPACE \
    --wait
```

#### Step 9: Install Ingress Controller. 

Follow this [link](https://learn.microsoft.com/en-us/azure/aks/ingress-basic?tabs=azure-cli#basic-configuration) for more details on setting up nginx ingress controller.

```bash
helm repo add ingress-nginx https://kubernetes.github.io/ingress-nginx
helm repo update

helm upgrade --install \
    ingress-nginx ingress-nginx/ingress-nginx \
  --namespace $DAPR_PERF_METRICS_NAMESPACE \
  --set controller.service.annotations."service\.beta\.kubernetes\.io/azure-load-balancer-health-probe-request-path"=/healthz
```

#### Step 10: Create Username and Password for Authentication

To create a basic authentication [username and password](https://kubernetes.github.io/ingress-nginx/examples/auth/basic/), use the following command, which will create an auth file and prompt you to provide a username and password.

```bash

> htpasswd -c auth <user name for prometheus pushgateway>

Reviewer: I think that this too could be a variable we set in step 3 but otherwise LGTM.
